sklearn/cluster/_k_means_fast.pyx

```diff
@@ -7,155 +7,256 @@
-        unsigned int n_features = centers.shape[1]
-        unsigned int n_samples = X.shape[0]
-        unsigned int store_distances = 0
-        unsigned int sample_idx, center_idx, feature_idx
-        unsigned int k
-        np.ndarray[floating, ndim=1] center_squared_norms
-        # the following variables are always double cause make them floating
-        # does not save any memory, but makes the code much bigger
-        DOUBLE inertia = 0.0
-        DOUBLE min_dist
-        DOUBLE dist
+        floating[::1] X_data = X.data
+        int[::1] X_indices = X.indices
+        int[::1] X_indptr = X.indptr
 
-    if floating is float:
-        center_squared_norms = np.zeros(n_clusters, dtype=np.float32)
-    else:
-        center_squared_norms = np.zeros(n_clusters, dtype=np.float64)
+        int n_samples = X.shape[0]
+        int n_features = X.shape[1]
+        int i, j
 
-    if n_samples == distances.shape[0]:
-        store_distances = 1
+        floating sq_dist = 0.0
+        floating inertia = 0.0
 
-    for center_idx in range(n_clusters):
-            center_squared_norms[center_idx] = _dot(
-                n_features, &centers[center_idx, 0], 1,
-                &centers[center_idx, 0], 1)
-
-    for sample_idx in range(n_samples):
-        min_dist = -1
-        for center_idx in range(n_clusters):
-            dist = 0.0
-            # hardcoded: minimize euclidean distance to cluster center:
-            # ||a - b||^2 = ||a||^2 + ||b||^2 -2 <a, b>
-            for k in range(X_indptr[sample_idx], X_indptr[sample_idx + 1]):
-                dist += centers[center_idx, X_indices[k]] * X_data[k]
-            dist *= -2
-            dist += center_squared_norms[center_idx]
-            dist += x_squared_norms[sample_idx]
-            dist *= sample_weight[sample_idx]
-            if min_dist == -1 or dist < min_dist:
-                min_dist = dist
-                labels[sample_idx] = center_idx
-                if store_distances:
-                    distances[sample_idx] = dist
-        inertia += min_dist
+        floating[::1] centers_squared_norms = row_norms(centers, squared=True)
+
+    for i in range(n_samples):
+        j = labels[i]
+        sq_dist = _euclidean_sparse_dense(
+            X_data[X_indptr[i]: X_indptr[i + 1]],
+            X_indices[X_indptr[i]: X_indptr[i + 1]],
+            centers[j], centers_squared_norms[j], True)
+        inertia += sq_dist * sample_weight[i]
 
     return inertia
 
 
+cdef void _relocate_empty_clusters_dense(np.ndarray[floating, ndim=2, mode='c'] X,
+                                         floating[::1] sample_weight,
+                                         floating[:, ::1] centers_old,
+                                         floating[:, ::1] centers_new,
+                                         floating[::1] weight_in_clusters,
+                                         int[::1] labels):
+    """Relocate centers which have no sample assigned to them."""
```

Comment:
I already added tests for the relocation of clusters (it's in master). It's does not only test this function isolated but kmeans globally. However I prepared a dataset for the test to make sure a cluster needs to be relocated and computed by hand the expected result. Don't you think such a test is enough ? (it's `test_relocated_clusters`)

Reply:
I think it would be useful to more finally test the function in isolation as I described above as a unit test rather than as an integration test for which we cannot make assertions on what happens at each step.

Reply:
done